Add check if importType succeds

JGiola opened the pull request:
SwiftDeclConverter::importSwiftNewtype now check if the storedUnderlyingType is null and return a nullptr to drop the typedef.

Resolves [SR-5614](https://bugs.swift.org/browse/SR-5614).

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4970,6 +4970,9 @@ SwiftDeclConverter::importSwiftNewtype(const clang::TypedefNameDecl *decl,
       decl->getUnderlyingType(), ImportTypeKind::Value, isInSystemModule(dc),
       Bridgeability::None, OTK_None);
 
+  if (storedUnderlyingType.isNull())
```
Nitpick: Type has an overload of `!` that can be used for this.

JGiola replied:
No problem, I've used `isNull()` only because I've found it used elsewhere and I thought it is the preferred way to use, I'm fixing right away

jrose-apple commented on test/ClangImporter/bad-ns-extensible-string-enum.swift:
```diff
@@ -0,0 +1,3 @@
+// RUN: %target-swift-frontend -typecheck %s -import-objc-header %S/Inputs/bad-ns-extensible-string-enum.h -verify
```
Two requests:
- Please add `// REQUIRES: objc_interop` to keep this from being run on non-ObjC-friendly platforms.
- Ideally, this would work with the "mock SDK" in test/Inputs/, which is a bit faster to build and doesn't depend on what's in the real SDK. We have a special syntax for that: `%target-swift-frontend(mock-sdk: %clang-importer-sdk)`. Does the test still pass with that?

JGiola replied:
I don't get it where I have to put the string in the second bullet point.

And one thing, there is somewhere come kind of guide on how to use the comments for the tests? For this I was merely copying from other files but I prefer to know what I am doing and avoid this pitfalls

jrose-apple replied:
Ah, right. There's some of it in [docs/Testing.md](https://github.com/apple/swift/blob/master/docs/Testing.md#writing-tests), but we should probably fill that out and organize it a little better. The format's the same as what LLVM uses, so the [LLVM Testing Guide](http://www.llvm.org/docs/TestingGuide.html#writing-new-regression-tests) is also a useful reference.

In this case, the string I put replaces your existing `%target-swift-frontend`.

JGiola replied:
I have changed the first line of the comment with `RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -typecheck -verify -I %S/Inputs/bad-ns-extensible-string-enum.h %s` but the compiler is crashing on me with "opening import file for module 'Swift': Not a directory"

I'm doing something wrong?

jrose-apple replied:
`-I` provides a search path, rather than importing the header as a bridging header. You're looking for `-import-objc-header`.

JGiola replied:
Now with this comment `RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -typecheck %s -import-objc-header %S/Inputs/bad-ns-extensible-string-enum.h -verify` it is failing with "failed to import bridging header"

jrose-apple replied:
Okay, that's better. I suspect what's failing is the use of `NS_EXTENSIBLE_STRING_ENUM`, which we probably don't define anywhere in test/Inputs/clang-importer-sdk/. We could put a definition in there, but for now you might as well replace it with the attribute it expands to, `__attribute__((swift_wrapper(struct)))`.

jrose-apple reviewed (commented):
Thanks for picking this up! A few comments.

jrose-apple commented:
Also, it'd be great to mention the SR in the commit message, so that we can find it again later if we need to.

JGiola commented:
My bad, I can push a rewrote commit with the changes or do you prefer it in a separate commit with the issue mentioned in that?

jrose-apple commented:
Rewrite is fine. No reason to preserve this particular history when people look at it in the future. Thanks!

JGiola committed:
Fix SR-5614, add check if importType succeds

SwiftDeclConverter::importSwiftNewtype now check if the
storedUnderlyingType is null and return a nullptr

JGiola commented:
Success! 🎉

jrose-apple reviewed (approved):
Looks great!

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Unrelated failure…but I guess we have to wait for it to be cleaned up.

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please test macOS
